x-pack/metricbeat/module/gcp/carbon/carbon.go

```diff
@@ -0,0 +1,320 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package carbon
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"time"
+
+	"cloud.google.com/go/bigquery"
+	"cloud.google.com/go/civil"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+
+	"github.com/elastic/beats/v7/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/x-pack/metricbeat/module/gcp"
+	"github.com/elastic/elastic-agent-libs/logp"
+	"github.com/elastic/elastic-agent-libs/mapstr"
+)
+
+const (
+	// metricsetName is the name of this metricset
+	metricsetName = "carbon"
+)
+
+// init registers the MetricSet with the central registry as soon as the program
+// starts. The New function will be called later to instantiate an instance of
+// the MetricSet for each host defined in the module's configuration. After the
+// MetricSet has been created then Fetch will begin to be called periodically.
+func init() {
+	mb.Registry.MustAddMetricSet(gcp.ModuleName, metricsetName, New)
+}
+
+// MetricSet holds any configuration or state information. It must implement
+// the mb.MetricSet interface. And this is best achieved by embedding
+// mb.BaseMetricSet because it implements all of the required mb.MetricSet
+// interface methods except for Fetch.
+type MetricSet struct {
+	mb.BaseMetricSet
+	config config
+	logger *logp.Logger
+}
+
+type config struct {
+	Period              time.Duration `config:"period" validate:"required"`
+	ProjectID           string        `config:"project_id" validate:"required"`
+	CredentialsFilePath string        `config:"credentials_file_path"`
+	CredentialsJSON     string        `config:"credentials_json"`
+	DatasetID           string        `config:"dataset_id" validate:"required"`
+	TableName           string        `config:"table_name"`
+}
+
+// Validate checks for deprecated config options
+func (c config) Validate() error {
+	if c.CredentialsFilePath == "" && c.CredentialsJSON == "" {
+		return errors.New("no credentials_file_path or credentials_json specified")
+	}
+
+	if c.Period.Hours() < 24 {
+		return fmt.Errorf("collection period for carbon footprint metricset %s cannot be less than 24 hours", c.Period)
```

Comment:
What do you think of logging the problem here but overriding the value with `24` and moving forward instead of triggering an error? (even not in this PR, as I suspect we use the same logic while validating other gcp settings)

Reply:
Yeah, I just reused this logic from the billing metricset. I'm fine doing that, but we should definitely do it for at least billing as well.

Reply:
We also can't override the value in `Validate()`, since we have a `config{}`, not a pointer to it.
Thinking about it, I don't know if this would be the best behavior. I'd be quite surprised and would probably expect a bug if I explicitly configured something to run hourly, and it only ran daily.

Reply:
This would not have been added to this PR anyway, so I'm resolving this conversation. We can continue later on :)